### PR TITLE
Add data loader for CDC

### LINKED ISSUE #300: Add data loader for CDC’s Socrata Data Portal
CDC is now publishing supply information (the same dataset that powers vaccines.gov) on their data portal at https://data.cdc.gov/Vaccinations/Vaccines-gov-COVID-19-vaccinating-provider-locatio/5jp2-pgaw. (This is a Socrata data portal, so can be accessed using Socrata’s standard API.)

This isn’t availability data, but does contain useful information about:
- What vaccine products are being supplied to a location (useful if we can’t get availability data about products from other sources).
- Other location-specific info, like location, position, hours, walk-ins allowed, etc.
- Vaccines.gov GUIDs that we can put in `external_ids`.

Each line is a unique location (vaccines.gov GUID) + vaccine product combination, so many locations will be listed more than once. Vaccine products are identified by their Unit-of-Use NDC10 code (the `ndc` column), which can be mapped to CVX codes and vaccine names with data from this CDC page: https://www.cdc.gov/vaccines/programs/iis/COVID-19-related-codes.html

This stuff is super useful (*especially* the product info), and we should write a loader that pulls it in regularly. It doesn’t update rapidly, so that loader should probably just run a couple times a day.

---

**This will also need some up-front work to map location GUIDs to our locations so we don’t create lots of duplicates the first time we run it.**
- The store number (`loc_store_no`) column might be useful for this, but requires having a way to identify which store brand that store number goes with.
- Vaccinate the States should also have concordances with these identifiers, and that might be a better/easier way to match things up initially.

### PR DESCRIPTION
**Pull request by astonm**
This PR is an approach to loading CDC data, and works with all of the states they support. For the sake of making sure we get good quality matches, it limits its current output to locations where we know the store number. Later on, we may be able to expand the number of locations loaded when we have more Vaccines.gov UUIDs in our system (perhaps from VtS data).

Fixes #300

### CODE REVIEW

**Comment by Mr0grog** [comment]
> I'm not sure what we're going to do about trying to match these locations to existing locations in our database

I haven’t read through the code here yet, but my thoughts here were:
- Start by only using locations that have the `loc_store_no` column filled in, and match those (requires parsing the `loc_name` column to figure out which store system it’s talking about, unfortunately).
- Match up other, non store-number locations out of band and add their vaccines.gov GUIDs as external IDs manually. After that’s done, turn this on for all rows, not just those with `loc_store_no`.

File: loader/src/sources/cdc/api.js
Comment (Mr0grog) [inline loader/src/sources/cdc/api.js:102]: I think it would be good to drop this line. The other loaders only have it for historic reasons, and it would probably be best to only set it in cases where we originally read in the value from our API.

File: loader/src/sources/cdc/api.js
Comment (Mr0grog) [inline loader/src/sources/cdc/api.js:108]: This should *probably* be based on `systemNameRe` instead, so we don’t flip stores from their name (e.g. `rite_aid` to `cdc` whenever this loader runs).

File: loader/src/sources/cdc/api.js
Comment (Mr0grog) [inline loader/src/sources/cdc/api.js:111]: Wondering if `titleCase()` would be useful here.

File: loader/src/sources/cdc/api.js
Comment (Mr0grog) [inline loader/src/sources/cdc/api.js:117]: It looks like there are a handful of records with no latitude/longitude. We should probably only set `position` if those are both present.

File: loader/src/sources/cdc/api.js
Comment (Mr0grog) [inline loader/src/sources/cdc/api.js:153]: Grrrr, looking at this again now, it’s not working well in the case where these values are VTrckS PINs instead of store numbers. (I didn’t even realize that was the case for some values in this column before!)

For VTrckS, the first 3 characters are an identifier for who issued the PIN, so in the comment above, `RA1` designates that this was issued by Rite Aid (the major pharmacies get to issue their own pins, then there are ones issued by CDC [not sure on the prefix there] and by state departments of health, e.g. NJ’s is `NJ<number>`). I have no idea where any official documentation for that is; I kept trying to find it but ultimately only gleaned this info through several conversations with CDC and DoH folks.

So anyway, the Rite Aid store number we should wind up with when handed `RA105587` is actually `5587` (so we also need to remove the leading zeroes to make it as normal as possible). (See our existing record for this example: https://getmyvax.org/api/edge/locations/rite_aid:5587)

File: loader/src/sources/cdc/api.js
Comment (Mr0grog) [inline loader/src/sources/cdc/api.js:216]: Not a big deal, but it *might* be worthwhile to move this NDC stuff up to `utils` or its own module or something.

File: loader/src/sources/cdc/api.js
Comment (Mr0grog) [inline loader/src/sources/cdc/api.js:128]: I’m not really sure we should set the `capacity` field… since this isn’t really appointments, we shouldn’t try to guess something here. Better to just set the top level `available` and `available_count` fields, I think.

**Review by Mr0grog — COMMENTED** [review]
The only real big deal here is the store number thing. I can’t believe they put VTrckS PINs in there sometimes and store numbers other times. ugh.

File: loader/src/sources/cdc/api.js
Comment (astonm) [reply loader/src/sources/cdc/api.js:8]: Oops! Thanks.

File: loader/src/sources/cdc/api.js
Comment (astonm) [reply loader/src/sources/cdc/api.js:111]: Yeah, checking the data set, there are plenty of city names that have wonky formatting, thanks.

File: loader/src/sources/cdc/api.js
Comment (astonm) [reply loader/src/sources/cdc/api.js:217]: I have some hesitation here just because I think this code is a bit hacky and not so robust to arbitrary NDC formats.

Comment (Mr0grog) [reply loader/src/sources/cdc/api.js:223]: That’s fair, and a good reason to keep it here for now.

File: loader/src/sources/cdc/api.js
Comment (astonm) [reply loader/src/sources/cdc/api.js]: Just took a stab at handling things both ways, using the guidelines you shared in Slack

> The format so far as I understand it (may not be 100% correct):
> - Always (?) 8 characters.
> - First 3 characters designates who issued the PIN (major pharmacies, states, CDC directly, not sure if there are other issuers/assigners/whatever)
> - States are <2-letter postal abbreviation><digit>
> - Pharmacies are <2-letter abbreviation>1 or <3-letter abbreviation>. The remaining characters are usually the store number, but are really up to the pharmacy as the issuer.
> - Remaining characters are up to the issuer and could be anything.

**Review by Mr0grog — COMMENTED** [review]
Ugh, sorry this just keeps going. A few minor notes inline and then…

I went ahead and checked to see how many locations this creates that we don’t already have and gut-check whether that seemed right:

- `walgreens`: 186

    This actually seem fine — I sampled a bunch and they were all either locations I didn’t correctly grab from our data (because they had funky provider names like `community_a_walgreens_pharmacy`) or are honestly missing from our data (they are listed in the Walgreens SMART SL API #135). The funky-named ones also have `walgreens:` external IDs, so they should match up fine in practice.

- `cvs`: 18

    Every single one of these seems to have no stock, and since we *should* already have a comprehensive listing of locations actually administering vaccines from the various CVS APIs, I’m thinking these might be places that have never been stocked and never administered. It sucks to hardcode things like this, but maybe we should include these properties on the 18 locations:

    ```js
    {
      is_private: true,
      internal_notes: 'Exists in CDC open data but not in CVS APIs; this location is probably not actually administering vaccines.'
    }
    ```

    For reference, the 18 locations were:

    ```
    cvs:9773
    cvs:9621
    cvs:9541
    cvs:3428
    cvs:5627
    cvs:5646
    cvs:4592
    cvs:8699
    cvs:8194
    cvs:8210
    cvs:8214
    cvs:8916
    cvs:739
    cvs:6006
    cvs:2972
    cvs:558
    cvs:1883
    cvs:7386
    ```

- `safeway`: 5

    These are all to do with the messiness of brand names. :(

    `safeway:3125` is `pak_n_save` in our DB (location ID: `7bd930d2-2bce-4b47-bcf9-366c35797b13`). It turns out that Pak-N-Sav is a sub-brand of Safeway, so we should probably “fix” this a) by giving it a `safeway:` external ID manually, and b) by [special-casing `pak_n_save` in the Vaccinespotter loader](https://github.com/usdigitalresponse/univaf/blob/16ae8b4153c0a22a5565e0503b15070a07f56ee2/loader/src/sources/vaccinespotter/index.js#L323-L341) so that it also creates `safeway:` external IDs. **(That should be a separate PR, and that OR the manual fix should be done before merging this PR.)**

    The rest are listed as Alberstons stores in our DB. Checking out other sources, it looks like these were probably recently rebranded (street view photos from 2017 show them as Albertsons-branded, but one with a very recent streetview shows Safeway, and the store listings on Albertsons’s website shows them as Safeways. We should probably just manually add `safeway:` external IDs for these in the database.

    - `safeway:414` is in our DB as `daeed1fd-ed16-4438-9a6c-b794f343a567`
    - `safeway:2248` is in our DB as `b70c960b-c6e5-4043-a1cd-c1a4e9cc5b28`
    - `safeway:469` is in our DB as `7d196a9e-c781-4fed-9705-22cebb3ec645`
    - `safeway:186` is in our DB as `a4ba3ad1-f746-4945-983e-99b8100c26de`

- `publix`: 15

    These all appear to be legitimately missing from our DB.

**Comment by Mr0grog** [comment]
The reverse (locations we’ve got but the CDC data set doesn’t) also shows some disparity we should investigate, but I don’t think is a blocker for this PR:

```js
{
  cvs: 24,
  walgreens: 295,
  kroger: 15,
  publix: 26,
  costco: 2,
  safeway: 5,
  rite_aid: 2
}
```

**Comment by Mr0grog** [comment]
Also, just realized I was probably not super clear last night on what I think needs to happen w/r/t the locations that looked problematic to me, so:

- Nothing to do for the `walgreens` and `publix` ones. The disparity looks correct, so this loader will do the right thing.
- The 5 `safeway` locations:
    - Need manual remediation in our DB before deploying this.
    - For the first one, the Vaccinespotter loader should also be amended to automatically set the right external ID and location name and provider, but as a separate PR from this.
    - Not sure there’s anything we can do in the Vaccinespotter loader for the other 4.
- The 18 CVS locations should probably be hard-coded to not be public. I’m not 100% on that, though, and am curious what you think.

File: loader/src/sources/cdc/api.js
Comment (jaronoff97) [inline loader/src/sources/cdc/api.js:17]: Never seen this generator syntax before, very cool!!

**Comment by astonm** [comment]
@Mr0grog thanks for all the feedback here. I trust your instincts on the CVS situation–seems like we can safely hide them for now.

File: loader/src/sources/cdc/api.js
Comment (Mr0grog) [inline loader/src/sources/cdc/api.js:272]: Very minor nit: everywhere else, we treat acronyms as words in camel case, so this should be `unexpectedCvses` or maybe `unexpectedCvsIds`.

File: loader/src/sources/cdc/api.js
Comment (Mr0grog) [inline loader/src/sources/cdc/api.js:213]: Saw the CVS one and realized I totally missed the same capitalization issue here.

File: loader/src/sources/cdc/api.js
Comment (Mr0grog) [inline loader/src/sources/cdc/api.js:300]: The `utils.oneLine` tag function is great for this. 🙂

**Review by Mr0grog — COMMENTED** [review]
Just style issues.

File: loader/src/sources/cdc/api.js
Comment (Mr0grog) [inline loader/src/sources/cdc/api.js:190]: Welp, I discovered some documentation we must have missed today that explains this, and the `supply_level` column is not an actual count in any way. It’s an enum describing *how long current supply should last*. https://data.cdc.gov/api/views/5jp2-pgaw/files/2a8f085d-04ab-4044-8148-3b7df4cb7088?download=true&filename=Outbound%20File%20Specification.pdf

| Value | Definition |
| ----: | ---------- |
| -1 | No report |
| 0 | No supply |
| 1 | < 24 hours of supply |
| 2 | 24–48 hours of supply |
| 3 | > 24 hours of supply |

…so we should probably only be using this to infer yes/no for `available`, and not try to set `available_count` with it at all.

Comment (astonm) [reply loader/src/sources/cdc/api.js:190]: 😮 

Will fix, thanks.

**Comment by astonm** [comment]
@Mr0grog I think this is already to after we update the database. Here's the SQL I have planned:

```
INSERT INTO external_ids (provider_location_id, system, value) VALUES 
 ('7bd930d2-2bce-4b47-bcf9-366c35797b13', 'safeway', '3125'),
 ('daeed1fd-ed16-4438-9a6c-b794f343a567', 'safeway', '414'),
 ('b70c960b-c6e5-4043-a1cd-c1a4e9cc5b28', 'safeway', '2248'),
 ('7d196a9e-c781-4fed-9705-22cebb3ec645', 'safeway', '469'),
 ('a4ba3ad1-f746-4945-983e-99b8100c26de', 'safeway', '186');
```

I'm thinking we just run it manually, but I can add a migration if you prefer.

**Comment by Mr0grog** [comment]
That looks good to me. I think running it manually makes sense and is a good way to go. 👍

**Review by Mr0grog — APPROVED** [review]
Looks good to me! Might be nice to have some simple tests, but this is probably also not worth holding up for that (we can do them in a second pass as we work to support for location types).

Feel free to merge after to running the SQL. 👍

**Comment by Mr0grog** [comment]
Oh, and I guess we'll need a quick follow-on to schedule this in Terraform. (1/day?)

**Pull request merged**
